File: android/app/src/main/kotlin/com/example/ble_proximity_bridge/DistanceEstimator.kt
package com.example.ble_proximity_bridge

import android.os.SystemClock

/**
 * Turns raw per-peer RSSI samples into rough distance estimates.
 *
 * Per peer: keep a sliding window of recent samples, take the median to
 * drop outliers, smooth it with a Kalman filter, then map it to a distance
 * bucket. The output is a bucket (0.5 / 1.5 / 3.5 / 8.0 m) on purpose:
 * RSSI jumps around too much for anything more precise.
 */
class DistanceEstimator {

    private val models = HashMap<Long, PeerModel>()

    @Synchronized
    fun clear() {
        models.clear()
    }

    /**
     * Adds one RSSI sample and returns the current estimate, or null until
     * there is enough data. One synchronized operation so BLE callbacks
     * cannot interleave.
     */
    @Synchronized
    fun addSample(peerId: Long, rssi: Double, senderDeviceKind: Int): Double? {
        if (rssi.isNaN()) return null
        val model = models.getOrPut(peerId) { PeerModel(senderDeviceKind) }
        return model.addAndEstimate(rssi)
    }

    /** Frees the model for a peer that left. */
    @Synchronized
    fun forget(peerId: Long) {
        models.remove(peerId)
    }

    private class PeerModel(senderDeviceKind: Int) {
        companion object {
            private const val MAX_SAMPLES = 20
            private const val WINDOW_MILLIS = 30_000L

            // Cutoffs depend on who sent the signal: iPhones usually read
            // hotter than Android phones at the same distance.
            private val RSSI_CUTOFFS_IOS_SENDER = doubleArrayOf(-55.0, -65.0, -75.0)
            private val RSSI_CUTOFFS_ANDROID_SENDER = doubleArrayOf(-70.0, -80.0, -90.0)

            private val BUCKETS_METERS = doubleArrayOf(0.5, 1.5, 3.5, 8.0)
        }

        private val cutoffs =
            if (senderDeviceKind == 0) RSSI_CUTOFFS_IOS_SENDER
            else RSSI_CUTOFFS_ANDROID_SENDER

        // (elapsedRealtime ms, rssi) pairs. elapsedRealtime is monotonic,
        // so clock changes cannot break the window.
        private val window = ArrayDeque<Pair<Long, Double>>()
        private val kalman = KalmanFilter1D(
            measurementNoise = 2.0,
            initialEstimateError = 2.0,
            processNoise = 0.05,
        )

        fun addAndEstimate(rssi: Double): Double? {
            val now = SystemClock.elapsedRealtime()
            window.addLast(now to rssi)
            while (window.size > MAX_SAMPLES) window.removeFirst()
            while (window.isNotEmpty() && now - window.first().first > WINDOW_MILLIS) {
                window.removeFirst()
            }
            if (window.isEmpty()) return null

            val median = medianOf(window.map { it.second })
            val smoothed = kalman.update(median)

            for (i in cutoffs.indices) {
                if (smoothed > cutoffs[i]) return BUCKETS_METERS[i]
            }
            return BUCKETS_METERS.last()
        }

        private fun medianOf(values: List<Double>): Double {
            val sorted = values.sorted()
            val middle = sorted.size / 2
            return if (sorted.size % 2 == 1) sorted[middle]
            else (sorted[middle - 1] + sorted[middle]) / 2.0
        }
    }
}

File: ios/Runner/DistanceEstimator.swift
import Foundation

/// Turns raw per-peer RSSI samples into rough distance estimates.
///
/// Per peer: keep a sliding window of recent samples, take the median to
/// kill outliers, smooth it with a Kalman filter, then map the result to a
/// distance bucket. The output is a bucket (0.5 / 1.5 / 3.5 / 8.0 m) on
/// purpose: RSSI jumps around several dB at a fixed distance, so anything
/// more precise would just be noise.
final class DistanceEstimator {

  private let lock = NSLock()
  private var models: [Int: PeerModel] = [:]

  /// Drops all per-peer history. Called when the bridge stops.
  func clear() {
    lock.lock()
    defer { lock.unlock() }
    models.removeAll()
  }

  /// Adds one RSSI sample and returns the current estimate, or nil until
  /// there's enough data. One locked operation so concurrent BLE callbacks
  /// can't interleave.
  func addSample(peerId: Int, rssi: Double, senderDeviceKind: Int) -> Double? {
    guard !rssi.isNaN else { return nil }
    lock.lock()
    defer { lock.unlock() }
    let model: PeerModel
    if let existing = models[peerId] {
      model = existing
    } else {
      model = PeerModel(senderDeviceKind: senderDeviceKind)
      models[peerId] = model
    }
    return model.addAndEstimate(rssi)
  }

  /// Frees the model for a peer that left.
  func forget(peerId: Int) {
    lock.lock()
    defer { lock.unlock() }
    models.removeValue(forKey: peerId)
  }

  private final class PeerModel {
    private static let maxSamples = 20
    private static let windowSeconds = 30.0

    // Cutoffs depend on who sent the signal: iPhones read hotter than most
    // Androids at the same distance.
    private static let rssiCutoffsIOSSender = [-55.0, -65.0, -75.0]
    private static let rssiCutoffsAndroidSender = [-70.0, -80.0, -90.0]

    private static let bucketsMeters = [0.5, 1.5, 3.5, 8.0]

    private let cutoffs: [Double]

    // (uptime seconds, rssi) pairs. systemUptime is monotonic, so clock
    // changes can't mess up the window.
    private var window: [(time: Double, rssi: Double)] = []
    private let kalman = KalmanFilter1D(
      measurementNoise: 2.0,
      initialEstimateError: 2.0,
      processNoise: 0.05
    )

    init(senderDeviceKind: Int) {
      cutoffs =
        senderDeviceKind == 0
        ? PeerModel.rssiCutoffsIOSSender
        : PeerModel.rssiCutoffsAndroidSender
    }

    func addAndEstimate(_ rssi: Double) -> Double? {
      let now = ProcessInfo.processInfo.systemUptime
      window.append((time: now, rssi: rssi))
      if window.count > PeerModel.maxSamples {
        window.removeFirst(window.count - PeerModel.maxSamples)
      }
      window.removeAll { now - $0.time > PeerModel.windowSeconds }
      guard !window.isEmpty else { return nil }

      let median = medianOf(window.map { $0.rssi })
      let smoothed = kalman.update(median)

      for (index, cutoff) in cutoffs.enumerated() where smoothed > cutoff {
        return PeerModel.bucketsMeters[index]
      }
      return PeerModel.bucketsMeters.last
    }

    private func medianOf(_ values: [Double]) -> Double {
      let sorted = values.sorted()
      let middle = sorted.count / 2
      if sorted.count % 2 == 1 {
        return sorted[middle]
      }
      return (sorted[middle - 1] + sorted[middle]) / 2.0
    }
  }
}
